### PR TITLE
assign seed and deterministic from config if they are None in args

### PR DESCRIPTION
**Pull request by AndreaPi**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Currently, as described in https://github.com/open-mmlab/mmdetection/issues/6553, it's not possible to set the random seed and the option Deterministic=True from the config file. This PR fixes that.

## Modification

Minor modifications to:

- `tools/train.py`
- `tools/analysis_tools/test_robustness.py`.

## BC-breaking (Optional)

None

### CODE REVIEW

File: tools/train.py [hunk 1]
```diff
@@ -144,6 +144,11 @@ def main():
     logger.info(f'Config:\n{cfg.pretty_text}')
 
     # set random seeds
+    if hasattr(cfg, "seed") and args.seed is None:
```
Comment (hhaAndroid) [inline tools/train.py:147]: If there are two seeds, we need to prompt the user to use `arg.seed`.

Comment (AndreaPi) [reply tools/train.py:147]: thanks @hhaAndroid . 

1. would the  following work?
```python
  if hasattr(cfg, "seed"):
        if args.seed is None:
            args.seed = cfg.seed
        else:
            logger.warning(f'Random seed passed both in config file and as an argument. \
            Please don''t set the seed in the config file if also passing it as an argument. \
            Using random seed passed as an argument : {args.seed}')
```
2. should I do the same for `args.determinism`?

Comment (AndreaPi) [reply tools/train.py:147]: @hhaAndroid could you please answer my questions? I'd be happy to close this PR as soon as possible

Comment (hhaAndroid) [reply tools/train.py:147]: It's best. And you can refer to
```python
    if hasattr(cfg, "seed") and args.seed is not None:
        seed = args.seed
        logger.warning..
    elif args.seed is not None:
        seed=args.seed
    elif hasattr(cfg, "seed"):
        seed=cfg.seed
    else:
        seed=None
```

Comment (hhaAndroid) [reply tools/train.py:147]: determinism can also be considered.

Comment (AndreaPi) [reply tools/train.py:147]: Thanks! I'm a bit busy today and tomorrow, but I'll try to get this done by Sunday.

Comment (hhaAndroid) [reply tools/train.py]: Ok

File: tools/analysis_tools/test_robustness.py [hunk 2]
```diff
@@ -207,8 +207,18 @@ def main():
...
+        print(f'Argument seed passed both in config file and as an argument. \
+            Please avoid setting the seed in the config file if also passing it as an argument. \
+            Using argument random seed {args.seed}')
+    elif args.seed is not None:
+        seed=args.seed
+    elif hasattr(cfg, "seed"):
+        seed=cfg.seed
+    else:
+        seed=None   
+    set_random_seed(seed)
```
Comment (hhaAndroid) [inline tools/analysis_tools/test_robustness.py:221]: miss `seed = init_random_seed(seed)`

Comment (AndreaPi) [reply tools/analysis_tools/test_robustness.py:222]: added, thanks

Comment (AndreaPi) [reply tools/analysis_tools/test_robustness.py:223]: @hhaAndroid @RangiLyu I fixed all the `flake8` issues, imported `init_random_seed` in `tools/analysis_tools/test_robustness.py ` (it was missing). Can you tell me what else is needed to close this PR? BTW, it appears that the linter checks are not repeated after I pushed the new commits, and thus they still show as failing. Should I close this PR and open a new one?

**Comment by RangiLyu** [comment]
Hi, please install pre-commit and then run `pre-commit run --all-files` to fix the lint.

File: tools/analysis_tools/test_robustness.py (hunk 2 above)
Comment (hhaAndroid) [reply tools/analysis_tools/test_robustness.py:223]: No need to close, just use this PR directly.

**Comment by AndreaPi** [comment]
> Hi, please install pre-commit and then run `pre-commit run --all-files` to fix the lint.

I actually installed `flake8` in VS Code, so the flake8 checks all pass now. Anyway, I can install `pre-commit` if it's strictly necessary. I run in a Docker container where I'm root. What should I do exactly? Are these commands enough?

```bash
pip install pre-commit
cd /mmdetection 
pre-commit install
pre-commit run --all-files
```

**Comment by AndreaPi** [comment]
@hhaAndroid @RangiLyu I installed `pre-commit` and ran `pre-commit run --all-files` but I get this error:

```bash
...# pre-commit run --all-files
[INFO] Initializing environment for https://gitlab.com/pycqa/flake8.git.
[INFO] Initializing environment for https://github.com/asottile/seed-isort-config.
[INFO] Initializing environment for https://github.com/timothycrosley/isort.
[INFO] Initializing environment for https://github.com/pre-commit/mirrors-yapf.
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/jumanjihouse/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/codespell-project/codespell.
[INFO] Initializing environment for https://github.com/myint/docformatter.
[INFO] Installing environment for https://gitlab.com/pycqa/flake8.git.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/asottile/seed-isort-config.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/timothycrosley/isort.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/mirrors-yapf.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/jumanjihouse/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('gem', 'build', 'fake_gem__.gemspec')
return code: 1
expected return code: 0
stdout:
    Executable `gem` not found
stderr: (none)
Check the log at /root/.cache/pre-commit/pre-commit.log
```

What should I do?  Here's the log content:

```bash
...# cat /root/.cache/pre-commit/pre-commit.log
### version information

```
pre-commit version: 2.16.0
sys.version:
    3.7.7 (default, May  7 2020, 21:25:33) 
    [GCC 7.3.0]
sys.executable: /opt/conda/bin/python
os.name: posix
sys.platform: linux
```

### error information

```
An unexpected error has occurred: CalledProcessError: command: ('gem', 'build', 'fake_gem__.gemspec')
return code: 1
expected return code: 0
stdout:
    Executable `gem` not found
stderr: (none)
```

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/pre_commit/error_handler.py", line 65, in error_handler
    yield
  File "/opt/conda/lib/python3.7/site-packages/pre_commit/main.py", line 396, in main
    return run(args.config, store, args)
  File "/opt/conda/lib/python3.7/site-packages/pre_commit/commands/run.py", line 417, in run
    install_hook_envs(to_install, store)
  File "/opt/conda/lib/python3.7/site-packages/pre_commit/repository.py", line 224, in install_hook_envs
    _hook_install(hook)
  File "/opt/conda/lib/python3.7/site-packages/pre_commit/repository.py", line 83, in _hook_install
    hook.prefix, hook.language_version, hook.additional_dependencies,
  File "/opt/conda/lib/python3.7/site-packages/pre_commit/languages/ruby.py", line 133, in install_environment
    prefix, ('gem', 'build', *prefix.star('.gemspec')),
  File "/opt/conda/lib/python3.7/site-packages/pre_commit/languages/helpers.py", line 52, in run_setup_cmd
    cmd_output_b(*cmd, cwd=prefix.prefix_dir, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pre_commit/util.py", line 154, in cmd_output_b
    raise CalledProcessError(returncode, cmd, retcode, stdout_b, stderr_b)
pre_commit.util.CalledProcessError: command: ('gem', 'build', 'fake_gem__.gemspec')
return code: 1
expected return code: 0
stdout:
    Executable `gem` not found
stderr: (none)
```

**Comment by AndreaPi** [comment]
@RangiLyu @hhaAndroid I haven't received any answers. `pre-commit run --all-files` fails (see error message above) and I don't know how to fix the error. Can you help me close this PR?

**Comment by hhaAndroid** [comment]
@AndreaPi Please sync the latest code.

**Comment by AndreaPi** [comment]
> @AndreaPi Please sync the latest code.

why should I? I'll just have the same issue as before (`pre-commit run --all-files` fails ), for which I didn't receive any support. Rather than wasting even more time on this PR, I'll just close it.

**Comment by AndreaPi** [comment]
> @AndreaPi Please sync the latest code.

I prefer to just close the PR. Syncing the latest code wouldn't solve my issues with `pre-commit` and since I didn't receive support on fixing them, I'll just give up on trying to get this PR merged.

**Pull request closed without merging**